### PR TITLE
perf: optimize canvas renderer performance

### PR DESCRIPTION
**Pull request by wang1212**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / Document optimization
- [ ] TypeScript definition update
- [ ] Refactoring
- [x] Performance improvement
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

see #1712 

### 💡 Background and solution

Currently, there is not much optimization in the core process of G's canvas renderer, resulting in poor performance under large data volumes. The following strategies can be introduced to optimize performance:

- [x] The forEach recursive call loop in the core code is refactored into a while loop to avoid stack overflow errors
- [x] Avoid repeated transformation matrix calculations and canvas context matrix updates during rendering
- [x] Avoid frequent `save()` and `restore()` calls to the canvas context state during the rendering process
- [ ] Support batch rendering

> Since changes in the core program flow may cause unexpected results, the performance optimization mode can be enabled through a flag in the early stage to avoid affecting existing products.
>```typescript
>const canvasRenderer = new CanvasRenderer({
>    enableRenderingOptimization: true,
>});
>```

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | perf: optimize canvas renderer performance |
| 🇨🇳 Chinese | perf: 优化 canvas 渲染器性能 |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: packages/g-lite/src/types.ts [hunk 1]
```diff
@@ -430,6 +430,13 @@ export interface RendererConfig {
    */
   enableSizeAttenuation: boolean;
 
+  /**
+   * Enable rendering optimization
```
Comment (hustcc) [inline packages/g-lite/src/types.ts:434]: 可以加一些更具体的描述，是什么优化的开关？

File: packages/g-plugin-canvas-renderer/src/shapes/styles/OptimizedDefault.ts [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+    context: CanvasRenderingContext2D,
+    parsedStyle: ParsedBaseStyleProps,
+    object: DisplayObject,
+    canvasContext: CanvasContext,
+    plugin: CanvasRendererPlugin,
+    runtime: GlobalRuntime,
+  ) {}
+
+  // #region common style
+  #applyCommonStyleToContext(
```
Comment (hustcc) [inline packages/g-plugin-canvas-renderer/src/shapes/styles/OptimizedDefault.ts:99]: 用 private 吧， # 需要上层业务编辑的时候，额外加一些配置吧？

**Comment by hustcc** [comment]
这个开启后有多少的优化效果？

**Comment by wang1212** [comment]
> 这个开启后有多少的优化效果？

在我的设备上测试：

- case1：1w 随机移动的矩形  fps 60 -> 95
- case2：2w 随机移动的矩形  fps 30 -> 45

发个版本后可以先验证下 s2 和 g6 的效果

File: packages/g-lite/src/services/RenderingService.ts [hunk 3]
```diff
@@ -220,52 +223,62 @@ export class RenderingService {
     canvasConfig: Partial<CanvasConfig>,
     renderingContext: RenderingContext,
   ) {
+    const self = this;
```
Comment (Aarebecca) [inline packages/g-lite/src/services/RenderingService.ts:226]: 这个目的是？

File: packages/g-plugin-canvas-renderer/src/CanvasRendererPlugin.ts [hunk 4]
```diff
@@ -219,8 +268,17 @@ export class CanvasRendererPlugin implements RenderingPlugin {
       mat4.multiply(this.vpMatrix, this.dprMatrix, camera.getOrthoMatrix());
 
       if (this.clearFullScreen) {
-        // console.log('canvas renderer fcp...', renderingContext.root.childNodes);
-        renderByZIndex(renderingContext.root, context);
+        // console.time('renderByZIndex');
+        if (enableRenderingOptimization) {
```
Comment (Aarebecca) [inline packages/g-plugin-canvas-renderer/src/CanvasRendererPlugin.ts:272]: 这里啥情况

Comment (wang1212) [reply packages/g-plugin-canvas-renderer/src/CanvasRendererPlugin.ts:272]: 优化不是默认开启的

image

**Review by Aarebecca — APPROVED** [review]

**Pull request merged**